### DISCUSSION TITLE
demonstrating {.content-visible when-format='html'} fails in posit.cloud

### DISCUSSION THREAD

**Discussion by rainer-rq-koelle** [quarto-dev/quarto-cli#9798] [Q&A] [answered]
### Description

Dear all, we are currently showcasing (and demonstrate) the benefits of a R/Quarto based workflow for technical documentation at work. Today's Q&A made us showcasing how to hide certain outputs in static formats.
However, it appears that the div `::: {.content-visible when-format='html'}` some-html-stuff-here :::` works flawlessly in the desktop app, but fails on posit.cloud.

Is there a chance to make this work within the posit.cloud environment?

Thanks for any pointers.

````qmd
---
title: "Demo Study"
format:
  docx: default
  #html: defailt
---

```{r}
#| label: setup
library(tidyverse)
library(plotly)
```

## Kicking-off the document

Very important text.

## Towards adding interactive plots

```{r}
gg <- starwars |> 
    select(name, height, mass, sex, gender, homeworld) |> 
    mutate(height_m = height/100, BMI = mass / height_m^2) |> 
    summarise(AVG_BMI = mean(BMI, na.rm = TRUE), .by = sex) |> 
    
    ggplot() + geom_col(aes(x = AVG_BMI, y = sex))
```

```{r}
gg + labs(title = "static")
```

Can we add/hide portion of the document dependent on the output format?

::: {.content-visible when-format="html"}

Will only appear in HTML.

```{r}
# we use plotly:: notation to highlight the package (which is already loaded above)
plotly::ggplotly(gg + labs(title = "interactive"))
```
:::

````

**Comment by cderv** [comment, marked as answer]
> However, it appears that the div ::: {.content-visible when-format='html'} some-html-stuff-here :::` works flawlessly in the desktop app, but fails on posit.cloud.

How does it fails on posit.cloud ? Do you have an error message ? 

If I am guessing correctly, you have a knitr rendering error mentioning non-HTML format and saying **webshot** or **webshot2** needs to be applied to get a screenshot to be included in non HTML format. You would get the same message in R Markdown ecosystem. 

So if I am right you need to either install the package webshot2 or webshot and its dependencies. Then it should work. I believe it is not working posit.cloud because the environment comes fresh, and you may not have installed those packages yet, while locally you may have them already. 

The reason of this message is because `.content-visible` is not applying before rendering time. It is only use to conditionnaly include markdown content in the output. It will not prevent a cell to be computed. And so this code should work accross format. 

There are ways to skip the rendering using **knitr** helpers (like `knitr::pandoc_to("docx")` to check if the current rendering format is docx). You could use that in an if for example. 

Hope it helps. 

i'll wait for you to confirm before closing

**Reply by rainer-rq-koelle** [reply]
@cderv thanks a lot for pointing out the obvious. I in fact had `{webshot2}` installed on the desktop computer, but not yet in the posit.cloud project. I probably focussed too much on the desired behaviour (i.e. when I change to a static format -- on the desktop -- the docx/pdf in rendering uneventful) rather than focussing on the command itself.
In fact I made the mistake to assume that `content-visible when-format='html'` would be *ignored* when rendering a static output format ... and would prevent that the div-ed portion would not be executed.

Installing `{webshot2}` on posit.cloud fixed the issue (or emulated the laptop environment). 

While this does the "demo"-bit and helps to showcase how we can combine static and interactive elements (e.g. ggplot2 in static output, and plotly in html output), the recipe is not correct. 
The correct(er) implementation should make use of the bespoke helper to control the sequence of execution.

Thanks for waiting before closing. This could be labelled as "user misperception" rather than an issue ;)
If I ever get an example going, I can append this to the closed issue.
